### ISSUE TITLE
lient should avoid using hardcoded types 

### ISSUE THREAD

**Issue by chenchix** [drasko/edgex-export#19]

**Comment by manuIO** [comment]
I agree 👍

**Comment by drasko** [comment]
@chenchix sure, these should be shared and that is why we put them in the common files, and not in `client` directory.

If I understand well you did not like the way we defined enums: https://github.com/drasko/edgex-export/blob/master/registration.go, i.e. as `iota` enums which is the way Go treats enums, but you decided to use whole strings.

Why should we use strings for these identifiers? They take more place and string comparison is more complex than simple verification of integer constant value.

@mijicd can you please give an opinion on this - I believe that in this case Go enums should be of type `int` and that `iota` should be used.

**Comment by feclare** [comment]
@drasko 
The problem is that is that the struct is used in the marshal from string to struct. And that fields are defined as strings in the protocol (https://wiki.edgexfoundry.org/download/attachments/328056/export-client.raml?version=1&modificationDate=1496455946000&api=v2).

I agree with you with the enum being more efficient than strings, but strings are a simpler solution and we can refactor it later if we found performance issues because of this.

**Comment by drasko** [comment]
@feclare I am aware that curent Java implementation uses strings. But I would like that we use this opportunity of rewrite and address the issues correctly - i.e. solve them how they should be solved and not give up on the quality because Java PoC did (probably) bad choice.

So, IMHO this should be addressed on Java side. You should open the issue for Tyler and Jim and we should discuss should we keep strings.

I am not worried only about optimization (it will be fast anyway) but the overall code style and approach is not quite tight. I see no technical justification in comparing strings instead simple integer flags.

I would like to understand also - why do you think that "strings are simpler", as you mentioned? Simpler in this case or in general? Because I think the opposite - int flags are simpler.

**Comment by feclare** [comment]
@drasko 
I understand this go microservices just as a reimplementation, and that it is important that we can test them with current java microservices. So they should be implemented according the current protocol.

It isn't that we can't improve the protocol and design, but that should be done, as you said in tickets of the main repos or mailing lists.

My comment about the strings was in the scenario of having the protocol with strings and an internal representation with integers. But I don't think you were suggesting that design.

**Comment by manuIO** [comment]
@drasko we have to use strings here: https://github.com/drasko/edgex-export/blob/master/client/registration.go#L56
If we use enums and want to read or write the correspoding string we need hardcoded values as we do in client or search in a table where are couples of enum - string. If we use strings they will be stored without any operation in DB, you have nothing to compare, plus is human readable.

**Comment by drasko** [comment]
@manuIO I think that using well named variables - like `CompGzip` are very much "human readable" in the code.

It is true though that if you store strings in database, it will be easier to inspect DB and conclude compression, encryption, etc without need for manual - but DB is not mean for humans to read it. It is meant for applications to consume the content.

Why do we need to use strings here: https://github.com/drasko/edgex-export/blob/master/client/registration.go#L56 - why don't you use Go enums we defined? If Java code does it wrong, we are not obliged to repeat the same errors. Rather - open the issues to Java people, and show them how this should be done correctly.

**Comment by drasko** [comment]
@feclare yes - exectly my point - improvements must be done though tickets to Java repos. And? Is this a problem?

This is the correct way to solve the problem - change the Java services do to the sane thing, do not adapt Go code to bad choices done in PoC. There is lot of "not so good" coices done in Java implementation - you see this with tight coupling, with ZMQ serialization, etc... The whole point of this rewrite is that we address these issues. Some of them are more fundamental (like ZMQ serialization), some are more benign (like this string comparison). But I think we should address them.

I do not believe that we can make Go interoperability with current Java version anyway (because some if the fundamental issues we seen). Look at this ZMQ issue - probably whole ZMQ will have to be replaced, as otherwise it will create the problems fro ARM compilation of C bindings. And this will demand new Java versions, which we can use to change these string flags as well, and some other issues that we find during the process...

This is of course my $0.2, but if majority wants to proceed with strings - let's go that way.

I want to upload this repo to EdgeX official GitHub org anyway, so that we can have several maintainers an vote on PRs.

**Comment by feclare** [comment]
we will make a PR with our changes, including this. Please do not change it yet.

**Comment by drasko** [comment]
@feclare sure, I do not even have the permissions to create EdgeX repos. This is something e have to see with Jim and @trcox.

**Comment by feclare** [comment]
@drasko I will make the pr to your repository, at least to receive comments about our design choices.

**Comment by drasko** [comment]
Yes, thanks. This is what I think as the best approach, we consolidate here the code a bit, and then we push all to official EdgeX repo next week for example, and take it from there.

**Comment by manuIO** [comment]
@drasko what else can we do here:  https://github.com/drasko/edgex-export/blob/master/client/registration.go#L56 ?

**Comment by drasko** [comment]
@manuIO this function does not make much sense to me. I do not see why is exactly needed, like this data is not known in advance. You see that everything is constant here, no dynamically created data or read from DB.

Whoever ask for this data should just have this information internally (like compression variants, etc...) in it's local code - because this is what our system uses, and it should be in the docs.

So, I do not quite get purpose of this function...

**Comment by manuIO** [comment]
@drasko if we remove this endpoint I agree 👍

**Comment by feclare** [comment]
I'm going to ask in the dev mail list, but: Is there any formal specs for the microservices api?

**Comment by drasko** [comment]
@feclare we should look at the wiki, I think I saw it somewhere. But we need our Swagger here, I think.

**Comment by drasko** [comment]
@manuIO for deleting/adding API endpoints you must sync to the Java guys and whole EdgeX devs. This is done via various channels, mostly ML and TSC meetings. Can you please take this obligation and send the appropriate mails with questions on ML?

**Comment by manuIO** [comment]
@drasko ok I'll

**Comment by drasko** [comment]
Need to be resoved for https://github.com/drasko/edgex-export/pull/20

**Comment by trcox** [comment]
> @feclare sure, I do not even have the permissions to create EdgeX repos. This is something e have to see with Jim and @trcox.

LF (@JPWKU) has the sole ability to open new repos in the edgexfoundry organization.

Good work here, tried to drop my input on all issues opened here.

Agreed, this is the time to find issues/coupling/better ways to handle things. API changes will have to be coordinated, but if there are better ways to handle things we are able to upstream changes to the Java side. Don't take the existing implementation as entirely set in stone beyond principles.

And certainly anything that blocks multi-language support should be identified and will be treated as a bug that we will resolve as a community.

**Comment by drasko** [comment]
@trcox thanks. I would propose that we start looking into this migration very soon. I think as soon as I pull the PR from @feclare and @chenchix we are ready to move to official EdgeX GitHub.

My biggest motivation for this move sooner rather than later is that issues are keep piling up. While it would be easy to transfer the code, we will loose issues and all these discussions we leas here, which are significant sources of development documentation.

**Comment by trcox** [comment]
Agreed @drasko, it has my vote as a top priority after the Barcelona Release is cut today.

**Comment by drasko** [comment]
Resolved by https://github.com/drasko/edgex-export/pull/20
